resteasy/src/main/java/com/linecorp/armeria/client/resteasy/ArmeriaJaxrsClientEngine.java

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.resteasy;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.io.ByteArrayOutputStream;
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.URI;
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.SSLContext;
+import javax.ws.rs.ProcessingException;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.client.Invocation;
+import javax.ws.rs.client.InvocationCallback;
+import javax.ws.rs.client.ResponseProcessingException;
+import javax.ws.rs.core.Response;
+
+import org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder;
+import org.jboss.resteasy.client.jaxrs.engines.AsyncClientHttpEngine;
+import org.jboss.resteasy.client.jaxrs.internal.ClientConfiguration;
+import org.jboss.resteasy.client.jaxrs.internal.ClientInvocation;
+import org.jboss.resteasy.client.jaxrs.internal.ClientResponse;
+
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpRequestWriter;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.RequestHeadersBuilder;
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.internal.common.resteasy.ByteBufferBackedOutputStream;
+import com.linecorp.armeria.internal.common.resteasy.HttpMessageStream;
+
+import io.netty.buffer.Unpooled;
+
+/**
+ * An implementation of {@link AsyncClientHttpEngine} based on Armeria {@link WebClient}.
+ * This provides the main entry point for JAX-RS client-side based on Armeria.
+ */
+@UnstableApi
+public class ArmeriaJaxrsClientEngine implements AsyncClientHttpEngine, Closeable {
```

Comment:
As mentioned in the JavaDoc, this ArmeriaResteasyClientBuilder builder is more like a convenience class. It should be still possible to construct the client using JAX-RS client API and Armeria engine directly as following:
```
     final Client jaxrsClient = ((ResteasyClientBuilder) ClientBuilder.newBuilder())
             .httpEngine(new ArmeriaJaxrsClientEngine(armeriaWebClient))
             .build();
```

Reply:
@trustin requested to add @UnstableApi to all public classes and interfaces, based on another PR feedback. I can remove this one, if you think it's not proper, but it would be nice to have more consistent requirements on @UnstableApi marking.

Reply:
Yes, I can also introduce `ArmeriaJaxrsClientEngineBuilder` to make it symmetric with `ResteasyService`/`ResteasyServiceBuilder`. In any case, this won't make it hidden, just like in case of `ResteasyService`. Do you agree?